cbits/win32/runProcess.c

```diff
@@ -0,0 +1,611 @@
+            *pHandleIn  = NULL;
+            *pHandleOut = NULL;
+            CloseHandle(hTemporaryIn);
+            CloseHandle(hTemporaryOut);
+            return FALSE;
+        }
+    }
+    *pHandleOut = hTemporaryOut;
+
+    return TRUE;
+}
+
+/*
+ * Function: mkNamedPipe
+ *
+ * Purpose:  create an named pipe with read and write ends being
+ *           optionally (non-)inheritable. Named pipes can ve read
+ *           asynchronously while anonymous pipes require blocking calls.
+ */
+static BOOL
+mkNamedPipe (HANDLE* pHandleIn, BOOL isInheritableIn,
+             HANDLE* pHandleOut, BOOL isInheritableOut)
+{
+#if 1 //defined(__IO_MANAGER_WINIO__)
+    HANDLE hTemporaryIn  = INVALID_HANDLE_VALUE;
+    HANDLE hTemporaryOut = INVALID_HANDLE_VALUE;
+    RPC_WSTR guidStr = NULL;
+    GUID guid;
+
+    /* First we create a new GUID to make the name of the pipe unique. Since
+       GUID are guaranteed to be unique system wide we don't need to retry.  */
+    ZeroMemory (&guid, sizeof (guid));
+    if (CoCreateGuid (&guid) != S_OK)
+        goto fail;
+
+    if (UuidToStringW ((UUID*)&guid, &guidStr) != S_OK)
+        goto fail;
+
+    /* Now we create the pipe name.  */
+    wchar_t pipeName[MAX_PATH];
+    if (-1 == swprintf_s (&pipeName[0],  MAX_PATH, L"\\\\.\\pipe\\haskell:process:%ls\n",  guidStr))
+        goto fail;
+
+    const int buffer_size = 8 * 1024;
+
+    RpcStringFreeW (&guidStr);
+
+    SECURITY_ATTRIBUTES secAttr;
+    ZeroMemory (&secAttr, sizeof(secAttr));
+    secAttr.nLength = sizeof(SECURITY_ATTRIBUTES);
+    secAttr.lpSecurityDescriptor = NULL;
+    secAttr.bInheritHandle = isInheritableIn;
+
+    /* Create one end of the pipe. Named pipes are a bit less secure than
+       anonymous pipes.  Because of this we restrict the pipe's access to only
+       one client and also only the local host.  This means after we create the
+       other end of the pipe it should be as secure as an anonymous pipe.  */
+    hTemporaryIn
+      = CreateNamedPipeW (&pipeName[0],
+                          PIPE_ACCESS_INBOUND | FILE_FLAG_OVERLAPPED | FILE_FLAG_FIRST_PIPE_INSTANCE,
+                          PIPE_TYPE_MESSAGE | PIPE_REJECT_REMOTE_CLIENTS,
+                          1, buffer_size, buffer_size,
+                          0,
+                          &secAttr);
+    if (hTemporaryIn == INVALID_HANDLE_VALUE)
+      goto fail;
+
+    /* And now create the other end using the inverse access permissions.  This
+       will give us the read and write ends of the pipe.  */
+    secAttr.bInheritHandle = isInheritableOut;
+    hTemporaryOut
+      = CreateFileW (&pipeName[0],
+                     GENERIC_WRITE,
+                     FILE_SHARE_WRITE,
+                     &secAttr,
+                     OPEN_EXISTING,
+                     FILE_FLAG_OVERLAPPED,
+                     NULL);
+    if (hTemporaryOut == INVALID_HANDLE_VALUE)
+      goto fail;
+
+    /* Set some optimization flags to makethe I/O manager operate more
```

Comment:
Missing space here.

Thanks for all of these comments, by the way. This makes it much easier to follow.